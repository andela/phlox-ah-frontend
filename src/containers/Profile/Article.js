import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

const Article = (props) => {
  return (
    <li className="article">
      <div className="photo">
        <img 
          src={props.article.imgUrl || props.placeholderImg } 
          alt="article-photo" 
        />
        <div className="title">
          {props.article.title}
        </div>
      </div>
      <div className="preview">
        {props.article.description}
      </div>
      <div className="footer">
        <span className="share">
          <a href="#">SHARE</a>
        </span>
        <span className="view">
          <a href="#">VIEW ARTICLE</a>
        </span>
        <span className="edit">
          <Link to={`/articles/${props.article.slug}/${props.article.status}/edit`}>
            <i className="fas fa-edit"></i>
          </Link>
        </span>
        <span className="delete">
          <button>
            <i className="fas fa-trash-alt"></i>
          </button>
        </span>
      </div>
    </li>
  )
}

Article.propTypes = {
  props: PropTypes.object,
};

export default Article;
