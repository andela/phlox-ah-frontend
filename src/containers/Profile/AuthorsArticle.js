import React, { Component } from 'react';
import { connect } from 'react-redux';
import './AuthorsArticle.scss';
import { getMyArticles } from '../BasePath';
import { Link } from 'react-router-dom';



class AuthorsArticle extends Component {

  componentDidMount() {
    this.props.getMyArticles();
  }

  editSelected(articleId) {
    console.log(articleId);
  }

  renderArticles(articles=[]) {

    const placeholderImg = "https://via.placeholder.com/300?text=AuthorsHaven";

    return articles.map(article => {
      return (
        <Article 
          placeholderImg={placeholderImg}
          key={ article.id } 
          article={ article } 
          {...this.props}
        />
      )
    });
  }

  render() {

    const { articles } = this.props;

    return (
      <div className="authors-article">
        <div className="heading">
          My Recent Articles
        </div>
        <ul className="articles-wrapper">
          {this.renderArticles(articles)}
        </ul>
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  articles: state.myArticle.articles
});

export default connect(mapStateToProps, {getMyArticles})(AuthorsArticle);


function Article (props){
  return (
    <li className="article">
      <div className="photo">
        <img 
          src={props.imgUrl || props.placeholderImg } 
          alt="article-photo" 
        />
        <div className="title">
          {props.article.title}
        </div>
      </div>
      <div className="preview">
        {props.article.description}
      </div>
      <div className="footer">
        <span className="share">
          <a href="#">SHARE</a>
        </span>
        <span className="view">
          <a href="#">VIEW ARTICLE</a>
        </span>
        <span className="edit">
          <Link to={`/articles/${props.article.slug}/slug`}>
            <i className="fas fa-edit"></i>
          </Link>
        </span>
        <span className="delete">
          <button>
            <i className="fas fa-trash-alt"></i>
          </button>
        </span>
      </div>
    </li>
  )
}
