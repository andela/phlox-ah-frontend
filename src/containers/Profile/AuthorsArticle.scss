div.authors-article {
  padding: 20px;
  width: 100%;
  
  div.heading {
    margin: 20px 0px;
    text-align: center;
    font-size: 25px;
    font-weight: 600;
    color: #1e5e6f;
  }

  ul.articles-wrapper {
    display: flex;
    flex-wrap: wrap;

    li.article {
      width: 361px;
      height: 260px;
      border: 1px solid #ccc;
      margin: 15px auto;
      box-shadow: 0 0 2px #ddd;
      background: #fff;

      a {
        color: inherit;
        display: block;
        width: 100%;
        height: 100%;
      }

      div.photo {
        height: 60%;
        width: 100%;
        overflow: hidden;
        position: relative;
      }

      div.photo img  {
        height: auto;
        width: 100%;
        object-fit: cover;
      }

      div.title {
        position: absolute;
        z-index: 10;
        bottom: 6px;
        font-size: 17px;
        font-weight: 600;
        letter-spacing: 0.5px;
        width: 100%;
        padding: 5px;
        color: #fff;
        text-shadow: 0 0 #444;
        text-align: center;
      }

      div.preview  {
        height: 25%;
        text-align: left;
        padding: 4px;
        font-size: 13px;
        text-align: justify;
        color: #555;
      }

      div.footer  {
        display: flex;
        justify-content: space-around;
        height: 15%;
        font-weight: 600;
        color: #2587a2;
        border-top: 1px solid #ccc;
        font-size: 14px;
        align-items: center;

        button {
          color: inherit;
          background: none;
          border: none;
          cursor: pointer;
        }
      }
    }

  }

}
