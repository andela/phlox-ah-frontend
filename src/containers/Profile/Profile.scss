body{
    background-color: #F8F9FA;
}
.profile-image {
    color: #06576D;
    width: 100px;
    height: 100px;
    border-radius: 50%;
    overflow: hidden;
    text-align: center;
    margin: auto;

    img {
        width: 100%;
        height: auto;
        border-radius: 0px;
    }
}
.container{
    background-color: #ffffff;
    width:90%;
    margin-top: 2px;
    min-height:800px;
}
.formCol{
    margin-top:100px;
    text-align: center;
    h5{
        font-weight: bold;
        margin-bottom:40px;
    }
    .updateButton{
        background-color: #06576D;
    }
}
@media only screen and (min-width: 993px) {
    .formCol{
        h5{
            font-weight: bold;
            margin-bottom:30px;
        }
    }
}