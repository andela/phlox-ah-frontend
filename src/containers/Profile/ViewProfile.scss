
$primary-color: #06576D;


@import '../../styles/variables';
div.profile {
  width: 100%;
  margin: 30px 0;

  a {
    text-decoration: none;
  }
}

div.profile div.profile-wrapper {
  margin: auto;
  width: 90%;
  display: flex;
  flex-wrap: wrap;
}

div.profile-wrapper div.profile-info,
div.profile-wrapper div.profile-edit {
  padding: 10px;
  background: rgba(0,0,0,0.04);
  margin: 5px;
}

div.profile-wrapper div.profile-info {
  width: 70%;
  display: flex;
  flex-wrap: wrap;
  min-width: 320px;
}

div.profile-info div.info-wrapper {
  display: inherit;
}

div.profile-info div.pi-photo {
  width: 100px;
}

div.pi-photo span.photo,
div.pi-photo span.change-photo {
  display: block;
}

div.pi-photo span.photo {
  border-radius: 50%;
  background: #ccc;
  width: 80px;
  height: 80px;
  margin: auto;
}

div.pi-photo span.change-photo {
  text-align: center;
  padding: 5px 0;
  display: none;
}

div.profile-info div.pi-detail {
  flex: 1;
  min-width: 300px;
  margin-left: 50px;
}

div.pi-detail div.username {
  font-size: 30px;
  font-weight: 600;
}

table.profile-data tr.data-row {
  border-bottom: none;
}

table.profile-data tr.data-row td:first-child{
  font-weight: 600;
  vertical-align: top;
}

table.profile-data tr.data-row td:nth-child(2){
  color: #777;
}

table.profile-data tr.data-row td{
  padding: 2px 3px;
  font-size: 16px;
  letter-spacing: 0.6px;
  line-height: 1.5;
}

div.profile-info div.followers {
  width: 100%;
  display: flex;
  flex-wrap: wrap;
  margin-top: 30px;
  padding-top: 6px;
}

div.followers div.follow,
div.followers div.following {
  flex-basis: 50%;
  padding: 5px;
  min-width: 300px;

  span.header {
    font-size: 20px;
    font-weight: 600;
    color: $primary-color;
    display: block;
    margin-left: 11px;
  }

  ul.authors {
    margin: 4px;
  }

  ul.authors li.author {
    display: flex;
    padding: 4px 0;
  }

  li.author a {
    display: inherit;
    text-decoration: none;
  }

  li.author div.photo {
    width: 50px;
    height: 50px;
    background: #ccc;
    border-radius: 50%;
    margin: 2px;
    margin-right: 5px;
  }

  li.author div.name {
    font-size: 16px;
    color: #777;
    line-height: 3;
    text-align: left;
    flex: 1;
    margin-left: 10px;
  }

  div.more {
    text-align: center;
    font-weight: 600;

    a {
      color: #039be5;
    }
  }
}

div.followers div.following {
  border-left: 1px solid #ccc;
  padding-left: 15px;
}

div.profile-wrapper div.profile-edit {
  flex: 1;
  min-width: 220px;
  align-self: self-start;

  div.edit-profile-link {
    text-align: center;
    font-size: 20px;
    font-weight: 600;
    text-decoration: underline;
    letter-spacing: 0.5px;
    color: #039be5;
    a {
      color: $themeColor;
      &:hover {
        color: darken($themeColor, 20);
        transition: 0.5s;
      }
    }
  }

  div.pref-heading {
    font-size: 18px;
    text-align: center;
    font-weight: 600;
    margin-top: 4px;
    color: #666;
  }

  div.tags ul.tags-wrapper {
    display: flex;
    flex-wrap: wrap;
    margin: 15px 0;
    max-height: 200px;
    overflow-y: auto;
  }

  ul.tags-wrapper li.tag {
    background: $primary-color;
    color: #fff;
    padding: 4px 13px;
    border-radius: 3px;
    font-size: 14px;
    letter-spacing: 0.5px;
    margin: 4px;
  }

  li.tag span.remove-tag {
    display: inline-block;
    margin-left: 8px;
    cursor: pointer;
  }
}


@media screen and (max-width: 890px) {
 div.profile-wrapper div.profile-info.tablet {
    width: 100%;
  }
}


@media screen and (max-width: 700px) {
  div.followers div.follow,
  div.followers div.following {
    margin: auto;
  }
  div.followers div.following {
    border: none;
    margin-top: 20px
  }
}


