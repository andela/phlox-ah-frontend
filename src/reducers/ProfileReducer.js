import { NEW_PROFILE, VIEW_PROFILE } from '../actionTypes/ProfileConstants';
import { asyncActionName } from '../util/AsyncUtil';

const initialState = {
  firstName: '',
  lastName: '',
  gender: '',
  bio: '',
  contact: '',
  profileImage: ''
};

const profileReducer = (state = initialState, action) => {
  switch (action.type) {
    case asyncActionName(NEW_PROFILE).success:
      return { ...action.payload.profile };
    case asyncActionName(VIEW_PROFILE).success:
      return { ...state, ...action.payload.profile.profile };
    default:
      return state;
  }
};

export default profileReducer;
