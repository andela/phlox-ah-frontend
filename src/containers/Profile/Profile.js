import React, { Component } from 'react';
import { connect } from 'react-redux';
import Avatar from 'react-avatar';
import PropTypes from 'prop-types';
import {
  Button, Input, Row, Col
} from 'react-materialize';
import { newProfile, viewProfile } from '../../requests/ProfileRequest';
import './Profile.scss';

/**
 *
 *
 * @class Profile
 * @extends {Component}
 */
class Profile extends Component {
  /**
   * @description - This method runs first in the class
   * @returns {object} profile
   * @memberof Profile
   */
  constructor() {
    super();
    this.state = {
      firstName: '',
      lastName: '',
      gender: '',
      bio: '',
      contact: '',
      profileImage: '',
      dataImage: ''
    };

    this.handleFileInputChange = this.handleFileInputChange.bind(this);
    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  /**
   * @description - This method runs after component has been mounted
   * @returns {object} profile
   * @memberof Profile
   */
  componentDidMount() {
    if (!this.props.profile.firstName) {
      this.props.viewProfile()
        .then(({ profile }) => this.updateState(profile));
    } else {
      this.updateState(this.props.profile);
    }
  }

  /**
   * @description - This method runs after component has been mounted
   * @returns {object} null
   * @param {object} data
   * @memberof Profile
   */
  updateState(data = {}) {
    this.setState({ ...data });
  }

  /**
   * @description - This method runs on input change
   * @param {object} e
   * @return {object} state
   * @memberof Profile
   */
  handleInputChange(e) {
    this.setState({
      [e.target.name]: e.target.value
    });
  }

  /**
   * @description - This method sets profile image
   * @param {object} e
   * @return {string} file url
   * @memberof Profile
   */
  handleFileInputChange(e) {
    const file = e.target.files[0];
    if (file.type === 'image/png' || file.type === 'image/jpg' || file.type === 'image/jpeg') {
      this.setState({
        profileImage: file
      });
      const reader = new FileReader();
      reader.onload = () => {
        const dataURL = reader.result;
        this.setState({
          dataImage: dataURL
        });
      };
      reader.readAsDataURL(file);
    }
  }


  /**
   * @description - This method handles the submit
   * @param {object} e
   * @return {object} new profile details
   * @memberof Profile
   */
  handleSubmit(e) {
    e.preventDefault();
    const formData = new FormData();
    Object.keys(this.state).forEach((key) => {
      formData.append(key, this.state[key]);
    });
    this.props.newProfile(formData, this.props);
  }

  /**
   *
   * @description - This method renders the jsx for this component
   * @returns {jsx} - jsx
   * @memberof Home
 */
  render() {
    return (
      <div className="container">
        <Row>
          <Col s={8} m={6} offset='s2 m3' className='grid-example formCol'>
          <h5>Edit My Profile</h5>
          <Row>
            <Col s={12}>
              <div className="profile-image">
                {
                  this.state.dataImage || this.props.profile.profileImage ? <img src={this.state.dataImage || this.props.profile.profileImage} alt="Profile Image" className="circle responsive-img"/>
                    : <Avatar name={this.props.user.username} size="75" round={true} />
                }
              </div>
            </Col>
            <Col s={12}><label htmlFor="profile-image" className="profile-image">Change</label></Col>
            <input type="file" onChange={this.handleFileInputChange} hidden id="profile-image" name="profileImage"/>
          </Row>
          <form onSubmit={this.handleSubmit}>
            <Row>
              <Input s={12} l={6} placeholder="First Name" label="Firstname" name="firstName" onChange={this.handleInputChange} value={this.state.firstName} />
              <Input s={12} l={6} placeholder="Last Name" label="Lastname" name="lastName" onChange={this.handleInputChange} value={this.state.lastName}/>
              <Input s={12} l={6} placeholder="Contact" label="Contact" name="contact" onChange={this.handleInputChange} value={this.state.contact}/>
              {/* <option value="" disabled selected>Choose gender</option> */}
              <div className="col input-field s12 l6" value={this.state.gender} onChange={this.handleInputChange}>
                <select className="browser-default s6" >
                  <option value="" disabled selected>Choose gender</option>
                  <option value="male">Male</option>
                  <option value="female">Female</option>
                </select>
              </div>
              <Input s={12} type='textarea' rows="4" placeholder="Bio" label="Bio" name="bio" onChange={this.handleInputChange} value={this.state.bio}/>
              <Col s={12}>
              <Button s={12} l={6}className="updateButton" waves='light'>Update Profile  <i className="fas fa-edit"></i></Button>
              </Col>
            </Row>
          </form>
          </Col>
        </Row>
      </div>
    );
  }
}
const mapStateToProps = state => ({
  profile: state.profile,
  user: state.user
});

Profile.propTypes = {
  newProfile: PropTypes.func,
  viewProfile: PropTypes.func,
  profile: PropTypes.object,
  user: PropTypes.object
};

export default connect(mapStateToProps, {
  newProfile,
  viewProfile
})(Profile);
