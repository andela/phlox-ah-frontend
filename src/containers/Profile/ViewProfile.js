/* eslint-disable no-script-url */
import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import Avatar from 'react-avatar';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { viewProfile } from '../../requests/ProfileRequest';
import FollowList from '../../components/FollowList/FollowList';
import Tags from '../../components/Tags/Tags';
import { getFollowings, getFollowers } from '../../requests/FollowRequests';
import AuthorsArticle from '../AuthorsArticle/AuthorsArticle';

import './ViewProfile.scss';
import './Profile.scss';

/**
 * @class ViewProfile
 * @extends {Component}
 */
class ViewProfile extends Component {
  /**
   * @description - This method runs after component has been mounted
   * @returns {object} articles
   * @memberof ViewProfile
   */
  componentDidMount() {
    this.props.getFollowings();
    this.props.getFollowers();
  }

  /**
   *
   * @description - This method renders the jsx for this component
   * @returns {jsx} - jsx
   * @memberof ViewProfile
   */
  render() {
    const tags = ['Religion', 'Sports', 'Technology', 'Music', 'Art', 'Software', 'Finance'];
    const listOfFollowings = (
      this.props.followings.slice(0, 3).map((following, i) => <FollowList
      key={i}>{following.Profile && following.Profile.profileImage ? <img src={following.Profile.profileImage} alt="Profile Image" className="circle responsive-img follow-pic"/>
        : <Avatar name={following.username} size="50" round={true} />}{following.username}</FollowList>));
    const listOfFollowers = (
      this.props.followers.slice(0, 3).map((followers, i) => <FollowList
      key={i}>{followers.Profile && followers.Profile.profileImage ? <img src={followers.Profile.profileImage} alt="Profile Image" className="circle responsive-img follow-pic"/>
        : <Avatar name={followers.username} size="50" round={true} />}{followers.username}</FollowList>));
    const tagList = tags.map((tag, i) => <Tags key={i}>{tag}</Tags>);

    return (
      <div className="profile">
        <div className="profile-wrapper">
          <div className="profile-info tablet">
            <div className="info-wrapper">
              <div className="pi-photo">
                <div className="profile-image">
                  {
                    this.props.profile.profileImage ? <img src={this.props.profile.profileImage} alt="Profile Image" className="circle responsive-img"/>
                      : <Avatar name={this.props.user.username} size="75" round={true} />
                  }
                </div>
              </div>
              <div className="pi-detail">
                <div className="username">
                  {`${this.props.profile.lastName} ${this.props.profile.firstName}`}
                </div>
                <table className="profile-data">
                  <tbody>
                    <tr className="data-row">
                      <td>Email: </td>
                      <td>{this.props.user.email}</td>
                    </tr>
                    <tr className="data-row">
                      <td>Bio: </td>
                      <td>
                        {this.props.profile.bio}
                      </td>
                    </tr>
                    <tr className="data-row">
                      <td>Contact: </td>
                      <td>
                        {this.props.profile.contact}
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>

            <div className="followers">
              <div className="follow">
                <span className="header">
                  Authors I Follow
                </span>
                <ul className="authors">
                  {listOfFollowings}
                  {!this.props.followings.length && <p>You are not following any author</p>}
                </ul>
                <div className="more">
                  <Link to="/following">View more</Link>
                </div>
              </div>
              {/* end of follow */}
              <div className="following">
                <span className="header">
                  Authors Following Me
                </span>
                <ul className="authors">
                  {listOfFollowers}
                  {!this.props.followers.length && <p>You do not have any followers</p>}
                </ul>
                 <div className="more">
                  <Link to="/followers">View more</Link>
                </div>
              </div>
              {/* end of following */}
            </div>
            {/* end of followers */}

            <AuthorsArticle {...this.props} />

          </div>
          {/* end of profile info */}
          <div className="profile-edit">
            <div className="edit-profile-link">
              <Link to='/profile/edit'>Edit Profile</Link>
            </div>
            <div className="pref-heading">
              My Article Preferences
            </div>
            <div className="tags">
              <ul className="tags-wrapper">
                {tagList}
              </ul>
            </div>
          </div>
        </div>


      </div>
    );
  }
}

const mapStateToProps = state => ({
  profile: state.profile,
  user: state.user,
  followings: state.followUser.followings,
  followers: state.followUser.followers
});

ViewProfile.propTypes = {
  viewProfile: PropTypes.func,
  firstName: PropTypes.string,
  lastName: PropTypes.string,
  gender: PropTypes.string,
  bio: PropTypes.string,
  contact: PropTypes.string,
  profileImage: PropTypes.string,
  profile: PropTypes.object,
  user: PropTypes.object,
  getFollowings: PropTypes.func.isRequired,
  getFollowers: PropTypes.func.isRequired,
  followings: PropTypes.array,
  followers: PropTypes.array,
};

export default connect(mapStateToProps, {
  viewProfile, getFollowings, getFollowers
})(ViewProfile);
