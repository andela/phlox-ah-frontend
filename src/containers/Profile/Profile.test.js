import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Profile from './Profile';
import ProfileReducer from '../../reducers/ProfileReducer';

const props = {
  profile: {
    firstName: '',
    lastName: '',
    gender: '',
    bio: '',
    contact: '',
    profileImage: ''
  },
  user: {
    username: '',
    email: ''
  },

};

const mockStore = configureMockStore([thunk]);
const store = mockStore({ ...props, viewProfile: jest.fn() });

let component;
let myComponent;


describe('<Profile/>', () => {
  beforeEach(() => {
    component = shallow(
      <Provider store={store}>
        <Profile />
      </Provider>
    );
    myComponent = component.dive({ context: { store } }).dive();
  });
  it('should render without throwing an error', () => {
    expect(component).toMatchSnapshot();
  });
  it('should have a Row component', () => {
    expect(myComponent.find('Row').exists()).toBe(true);
  });
  it('should have a form tag with class col', () => {
    expect(myComponent.find('form').exists()).toBe(true);
  });
  it('should have a Col component', () => {
    expect(myComponent.find('Col').exists()).toBe(true);
  });
  it('should have a Update profile button', () => {
    expect(myComponent.find('Button').exists()).toBe(true);
  });
  it('should have a input field', () => {
    expect(myComponent.find('Input').exists()).toBe(true);
  });

  it('should dispatch an action to create profile', () => {
    const payload = {
      profile: props.profile,
      message: 'Profile created successfully',
      success: true
    };
    const action = {
      type: 'NEW_PROFILE_SUCCESS',
      payload
    };
    const newState = ProfileReducer({}, action);
    expect(newState).toEqual({ ...props.profile });
  });
});
