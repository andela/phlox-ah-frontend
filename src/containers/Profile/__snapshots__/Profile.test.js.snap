// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Profile/> should render without throwing an error 1`] = `<Connect(Profile) />`;
